src/parquet/column/writer.cc

```diff
@@ -0,0 +1,174 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/column/writer.h"
+
+#include "parquet/encodings/plain-encoding.h"
+
+namespace parquet {
+
+// ----------------------------------------------------------------------
+// ColumnWriter
+
+ColumnWriter::ColumnWriter(const ColumnDescriptor* descr,
+    std::unique_ptr<PageWriter> pager, int64_t expected_rows,
+    MemoryAllocator* allocator):
+    descr_(descr), pager_(std::move(pager)), expected_rows_(expected_rows),
+    allocator_(allocator),
+    num_buffered_values_(0), num_buffered_encoded_values_(0),
+    num_rows_(0), total_bytes_written_(0) {
+  InitSinks();
+}
+
+void ColumnWriter::InitSinks() {
+  definition_levels_sink_ = std::unique_ptr<InMemoryOutputStream>(
+      new InMemoryOutputStream());
+  repetition_levels_sink_ = std::unique_ptr<InMemoryOutputStream>(
+      new InMemoryOutputStream());
+  values_sink_ = std::unique_ptr<InMemoryOutputStream>(
+      new InMemoryOutputStream());
+}
+
+void ColumnWriter::WriteDefinitionLevels(int64_t num_levels, int16_t* levels) {
+  definition_levels_sink_->Write(reinterpret_cast<uint8_t*>(levels),
+      sizeof(int16_t) * num_levels);
+}
+
+void ColumnWriter::WriteRepetitionLevels(int64_t num_levels, int16_t* levels) {
+  repetition_levels_sink_->Write(reinterpret_cast<uint8_t*>(levels),
+      sizeof(int16_t) * num_levels);
+}
+
+std::shared_ptr<Buffer> ColumnWriter::RleEncodeLevels(
+    const std::shared_ptr<Buffer>& buffer, int16_t max_level) {
+  // TODO: This only works with due to some RLE specifics
+  int64_t rle_size = 2 * num_buffered_values_ + sizeof(uint32_t);
+  std::shared_ptr<OwnedMutableBuffer> buffer_rle =
+    std::make_shared<OwnedMutableBuffer>(rle_size, allocator_);
+  level_encoder_.Init(Encoding::RLE, max_level,
+      num_buffered_values_, buffer_rle->mutable_data() + sizeof(uint32_t),
+      buffer_rle->size() - sizeof(uint32_t));
+  int encoded = level_encoder_.Encode(num_buffered_values_,
+      reinterpret_cast<const int16_t*>(buffer->data()));
+  DCHECK_EQ(encoded, num_buffered_values_);
+  reinterpret_cast<uint32_t*>(buffer_rle->mutable_data())[0] = level_encoder_.len();
+  buffer_rle->Resize(level_encoder_.len() + sizeof(uint32_t));
+  return std::static_pointer_cast<Buffer>(buffer_rle);
+}
+
+void ColumnWriter::WriteNewPage() {
+  // TODO: Currently we only support writing DataPages
+  std::shared_ptr<Buffer> definition_levels = definition_levels_sink_->GetBuffer();
+  std::shared_ptr<Buffer> repetition_levels = repetition_levels_sink_->GetBuffer();
+  std::shared_ptr<Buffer> values = values_sink_->GetBuffer();
```

Comment:
This may cause some allocation-deallocation thrashing versus reusing the same buffers after completing each data page. Perhaps the output stream can be changed to enable buffer reuse and resetting without deallocation.